hypervisor/debug/profiling.c

```diff
@@ -0,0 +1,1395 @@
+		offset = sbuf->size - sbuf->tail;
+		(void)memcpy_s(to, offset, data, offset);
+
+		/* 2nd part */
+		to = (void *)sbuf + SBUF_HEAD_SIZE;
+
+		if (size - offset > 0U) {
+			(void)memcpy_s(to, size - offset,
+				data + offset, size - offset);
+		}
+	} else {
+		dev_dbg(ACRN_DBG_PROFILING, "non-wrap-around!");
+		(void)memcpy_s(to, size, data, size);
+	}
+
+	sbuf->tail = next_tail;
+
+	return (int)size;
+}
+
+/*
+ * Read profiling data and transferred to SOS
+ * Drop transfer of profiling data if sbuf is full/insufficient and log it
+ */
+static int profiling_generate_data(int32_t collector, uint32_t type)
+{
+	uint32_t i, remaining_space = 0U;
+	int 	ret = 0;
+	struct data_header pkt_header;
+	uint64_t payload_size = 0UL;
+	void *payload = NULL;
+	struct shared_buf *sbuf = NULL;
+	struct sep_state *ss = &(get_cpu_var(profiling_info.sep_state));
+	struct sw_msr_op_info *sw_msrop
+		= &(get_cpu_var(profiling_info.sw_msr_op_info));
+
+	dev_dbg(ACRN_DBG_PROFILING, "%s: entering cpu%d",
+		__func__,  get_cpu_id());
+
+	if (collector == COLLECT_PROFILE_DATA) {
+		sbuf = (struct shared_buf *)
+				per_cpu(sbuf, get_cpu_id())[ACRN_SEP];
+
+		if (sbuf == NULL) {
+			ss->samples_dropped++;
+			dev_dbg(ACRN_DBG_PROFILING, "%s: sbuf is NULL exiting cpu%d",
+				__func__,  get_cpu_id());
+			return 0;
+		}
+
+		if (ss->pmu_state == PMU_RUNNING) {
+			if (sbuf->tail >= sbuf->head) {
+				remaining_space = sbuf->size
+						- (sbuf->tail - sbuf->head);
+			} else {
+				remaining_space = sbuf->head - sbuf->tail;
+			}
+
+			/* populate the data header */
+			pkt_header.tsc = rdtsc();
+			pkt_header.collector_id = collector;
+			pkt_header.cpu_id = get_cpu_id();
+			pkt_header.data_type = 1U << type;
+
+			switch (type) {
+			case CORE_PMU_SAMPLING:
+				payload_size = CORE_PMU_SAMPLE_SIZE;
+				payload = &get_cpu_var(profiling_info.pmu_sample);
+				break;
+			case LBR_PMU_SAMPLING:
+				payload_size = CORE_PMU_SAMPLE_SIZE
+					+ LBR_PMU_SAMPLE_SIZE;
+				payload = &get_cpu_var(profiling_info.pmu_sample);
+				break;
+			case VM_SWITCH_TRACING:
+				payload_size = VM_SWITCH_TRACE_SIZE;
+				payload = &get_cpu_var(profiling_info.vm_switch_trace);
+				break;
+			default:
+				pr_err("%s: unknown data type %u on cpu %d",
+					__func__, type, get_cpu_id());
+				ret = -1;
+				break;
+			}
+			if (ret == -1) {
+				return 0;
+			}
+			pkt_header.payload_size = payload_size;
+
+			if ((uint64_t)remaining_space < (DATA_HEADER_SIZE + payload_size)) {
+				ss->samples_dropped++;
+				dev_dbg(ACRN_DBG_PROFILING,
+				"%s: not enough space left in sbuf[%d: %d] exiting cpu%d",
+				__func__, remaining_space,
+				DATA_HEADER_SIZE + payload_size, get_cpu_id());
+				return 0;
+			}
+
+			for (i = 0U; i < (uint32_t)((DATA_HEADER_SIZE - 1U) / SEP_BUF_ENTRY_SIZE + 1U); i++) {
+				(void)sbuf_put((struct shared_buf *)sbuf,
+					(uint8_t *)&pkt_header + i * SEP_BUF_ENTRY_SIZE);
+			}
+
+			for (i = 0U; i < (uint32_t)((payload_size - 1U) / SEP_BUF_ENTRY_SIZE + 1U); i++) {
```

Comment:
define i as uint64_t , remove the (uint32_t) cast

Reply:
Done